oauth_provider/models/oauth_provider_scope.py

```diff
@@ -0,0 +1,338 @@
+
+        return records
+
+    @api.multi
+    def delete_record(self, records):
+        """ Delete a recordset that is within the current scope.
+
+        Args:
+            records (BaseModel): Recordset to delete.
+
+        Raises:
+            OauthScopeValidationException: If records are not within the
+                current scope.
+        """
+
+        if not self._validate_scope_record(records):
+            raise OauthScopeValidationException('record')
+
+        records.unlink()
+
+    @api.multi
+    def _get_filter_domain(self, eval_context, record_ids=None):
+        """ Return the scope's domain.
+
+        Args:
+            eval_context (dict): Base eval context, such as provided by
+                `ir_filter_eval_context`
+            record_ids (list[ints], optional): ID of record(s) to explicitly
+                query.
+
+        Returns:
+            list of tuples: Domain of the scope, in standard Odoo format.
+        """
+
+        self.ensure_one()
+
+        if record_ids is None:
+            record_ids = []
+
+        filter_domain = [(1, '=', 1)]
+
+        if self.filter_id:
+            filter_domain = safe_eval(
+                self.filter_id.sudo().domain,
+                eval_context,
+            )
+
+        if record_ids:
+            filter_domain.append(
+                ('id', 'in', record_ids),
+            )
+
+        return filter_domain
+
+    @api.multi
+    def _get_scoped_records(self, model_name, eval_context=None,
+                            add_domain=None, record_ids=None):
+        """ Return records that are within the scopes in the recordset.
+
+        Args:
+            model_name (str): Name of model to operate on.
+            eval_context (dict, optional): Base eval context, such as provided
+                by `ir_filter_eval_context`.
+            add_domain (list of tuples, optional): Domain to append to the
+                `filter_domain` that is defined in the scope.
+            record_ids (list[int], optional): ID of record to find. Will
+                only return this record, if defined.
+
+        Returns:
+            BaseModel: Recordset matching the scope.
+        """
+
+        self.ensure_one()
+        if eval_context is None:
+            eval_context = self.ir_filter_eval_context
+        if add_domain is None:
+            add_domain = []
+        filter_domain = self._get_filter_domain(eval_context, record_ids)
+        return self.env[model_name].search(filter_domain + add_domain)
+
+    @api.multi
+    def _validate_scope_record(self, records):
+        """ Validate that the recordset is within the current scope.
+
+        Args:
+            records (BaseModel): Recordset to validate.
+
+        Returns:
+            bool: Indicating whether the records are within scope.
+        """
+
+        scoped_records = self.env[records._name]
+        for scope in self:
+            scoped_records |= scope._get_scoped_records(
```

Comment:
I guess this will not work if you have scopes using different models on the same token.